### PR TITLE
Support for archiving to minio over ssl using custom CA

### PR DESCRIPTION
**Pull request by iamrz1**
xref:https://github.com/wal-g/wal-g/issues/250

### CODE REVIEW

**Comment by iamrz1** [comment]
@x4m , @g0djan PTAL

File: internal/storages/s3/session.go [hunk 1]
```diff
@@ -6,10 +6,15 @@ import (
 	"github.com/aws/aws-sdk-go/aws/session"
 	"github.com/aws/aws-sdk-go/service/s3"
 	"github.com/pkg/errors"
+	"os"
 	"strconv"
 	"strings"
 )
 
+const (
+	S3_CA_CERT_FILE = "WALG_S3_CA_CERT_FILE"
```
Comment (g0djan) [inline internal/storages/s3/session.go]: Add this key to internal/config.go in allowedConfigKeys

File: internal/storages/s3/session.go [hunk 2]
```diff
@@ -79,5 +84,15 @@ func createSession(bucket string, settings map[string]string) (*session.Session,
 	}
 	config = config.WithRegion(region)
 
+	caFilePath := os.Getenv(S3_CA_CERT_FILE)
+	if caFilePath != "" {
+		if file, err := os.Open(caFilePath); err == nil {
+			defer file.Close()
```
Comment (g0djan) [inline internal/storages/s3/session.go]: use loggedClose from utility

File: internal/storages/s3/session.go (hunk 1 above)
Comment (iamrz1) [reply internal/storages/s3/session.go]: Do we still use os.Getenv(S3_CA_CERT_FILE) ?

Comment (g0djan) [reply internal/storages/s3/session.go]: Try to use internal.GetSettingValue

Comment (iamrz1) [reply internal/storages/s3/session.go]: I initially tried to use `internal.GetSettingValue(S3_CA_CERT_FILE)` which required me to import `github.com/wal-g/wal-g/internal`, that in turn gave me an import cycle error during build:
```
import cycle not allowed
package github.com/wal-g/wal-g/main/pg
        imports github.com/wal-g/wal-g/cmd/pg
        imports github.com/wal-g/wal-g/internal
        imports github.com/wal-g/wal-g/internal/storages/s3
        imports github.com/wal-g/wal-g/internal
Makefile:16: recipe for target 'pg_build' failed
```
So I kept getEnv() for the time being.

File: internal/storages/s3/session.go [hunk 3]
```diff
@@ -96,3 +98,10 @@ func createSession(bucket string, settings map[string]string) (*session.Session,
 
 	return session.NewSession(config)
 }
+
+func loggedClose(c io.Closer) {
```
Comment (g0djan) [inline internal/storages/s3/session.go]: If you want to pass a message to LoggedCLose then change the signature of the function, don't copy it. https://en.wikipedia.org/wiki/Don%27t_repeat_yourself

File: internal/storages/s3/session.go (hunk 1 above)
Comment (iamrz1) [reply internal/storages/s3/session.go]: I initially tried to use `internal.GetSettingValue(S3_CA_CERT_FILE)` which required me to import `github.com/wal-g/wal-g/internal`, that in turn gave me an import cycle error during build:
```
import cycle not allowed
package github.com/wal-g/wal-g/main/pg
        imports github.com/wal-g/wal-g/cmd/pg
        imports github.com/wal-g/wal-g/internal
        imports github.com/wal-g/wal-g/internal/storages/s3
        imports github.com/wal-g/wal-g/internal
Makefile:16: recipe for target 'pg_build' failed
```
Basically internal/storage_adapter.go imports internal/storages/s3, and now I have to import internal in internal/storages/s3, which become a circular thing.
So I kept getEnv() for the time being.

Comment (iamrz1) [reply internal/storages/s3/session.go]: I found a better way: getting values from settingNames.

**Comment by iamrz1** [comment]
Hello @g0djan,
 I have updated the pull. Can you please have a look again?

**Comment by g0djan** [comment]
Yep. Now you can't import loggedClose from internal because it has cyclic error. Try to move utility to new package "utility" from "internal" and use loggedClose from this package instead of copy-paste.

**Comment by iamrz1** [comment]
> Yep. Now you can't import loggedClose from internal because it has cyclic error. Try to move utility to new package "utility" from "internal" and use loggedClose from this package instead of copy-paste.

Okay, I am moving the function to another folder (wal-g/utility) outside wal-g/internal.

**Comment by g0djan** [comment]
@iamrz1 I wait for the fix. I guess tests crashed because you update not all of the imports of the utility package

**Comment by iamrz1** [comment]
@g0djan, fixed. It was due to how test script copies a few select directories.

**Pull request merged**